File: src/nodejs/client/cec-client.js
#!/usr/bin/env node
'use strict';

/*
 * cec-client.js — an interactive test client for the libCEC Node.js binding.
 *
 * This is the Node.js counterpart of the C++ `cec-client` (src/cec-client): a
 * deliberately verbose REPL whose purpose is to *demonstrate* every part of the
 * binding's API and to *exercise* the library against real hardware. Clarity is
 * valued over brevity — each command handler is a small, self-contained example
 * of one API call, so this file doubles as living documentation.
 *
 * Usage:
 *   node client/cec-client.js [options] [port]
 *
 *   [port]                 the adapter to open (e.g. /dev/ttyACM0). Auto-detected
 *                          when omitted.
 *   -l, --list-devices     list the detected adapters and exit.
 *   -i, --info             print the libCEC build info and exit.
 *   -t, --type {p|r|t|a}   the device type to announce (playback/recording/
 *                          tuner/audio). Defaults to recording.
 *   -p, --port {n}         the HDMI port the adapter is connected to.
 *   -b, --base {addr}      the logical address the adapter is connected to.
 *   -d, --log-level {1-31} the initial log-level bitmask (see cectypes.h).
 *   -m, --monitor          start in monitor-only mode (claims no logical address).
 *   -h, --help             show this help and exit.
 *
 * Type `h` at the prompt for the interactive command list.
 */

const readline = require('readline');
const cec = require('..');

// ---------------------------------------------------------------------------
// argument parsing
// ---------------------------------------------------------------------------

function parseArgs(argv) {
  const opts = {
    port: null,
    deviceType: cec.CecDeviceType.RecordingDevice,
    hdmiPort: null,
    baseDevice: null,
    logMask: cec.CecLogLevel.Error | cec.CecLogLevel.Warning | cec.CecLogLevel.Notice,
    monitorOnly: false,
    listDevices: false,
    info: false,
    help: false,
  };
  const typeMap = {
    p: cec.CecDeviceType.PlaybackDevice,
    r: cec.CecDeviceType.RecordingDevice,
    t: cec.CecDeviceType.Tuner,
    a: cec.CecDeviceType.AudioSystem,
  };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    switch (a) {
      case '-l': case '--list-devices': opts.listDevices = true; break;
      case '-i': case '--info': opts.info = true; break;
      case '-h': case '--help': opts.help = true; break;
      case '-m': case '--monitor': opts.monitorOnly = true; break;
      case '-t': case '--type': opts.deviceType = typeMap[argv[++i]] ?? opts.deviceType; break;
      case '-p': case '--port': opts.hdmiPort = parseInt(argv[++i], 10); break;
      case '-b': case '--base': opts.baseDevice = parseInt(argv[++i], 10); break;
      case '-d': case '--log-level': opts.logMask = parseInt(argv[++i], 10); break;
      default:
        if (a.startsWith('-')) { console.error(`unknown option: ${a}`); process.exit(2); }
        opts.port = a; // a bare argument is the com port
    }
  }
  return opts;
}

// ---------------------------------------------------------------------------
// small formatting helpers
// ---------------------------------------------------------------------------

// libCEC reports every logical address as an enum; render it the way cec-client
// does, e.g. "TV (0)".
function fmtAddr(address) {
  return `${cec.logicalAddressToString(address)} (${address})`;
}

// a 16-bit physical address is a nibble per HDMI hop: 0x1234 -> "1.2.3.4".
function fmtPhysical(pa) {
  return [(pa >> 12) & 0xf, (pa >> 8) & 0xf, (pa >> 4) & 0xf, pa & 0xf].join('.');
}

// reverse-lookup a CecAdapterType value to its enum name (there is no native
// adapter-type-to-string helper).
function adapterTypeName(type) {
  const name = Object.keys(cec.CecAdapterType).find((k) => cec.CecAdapterType[k] === type);
  return name || `type ${type}`;
}

function fmtLogLevel(level) {
  switch (level) {
    case cec.CecLogLevel.Error: return 'ERROR';
    case cec.CecLogLevel.Warning: return 'WARNING';
    case cec.CecLogLevel.Notice: return 'NOTICE';
    case cec.CecLogLevel.Traffic: return 'TRAFFIC';
    case cec.CecLogLevel.Debug: return 'DEBUG';
    default: return `LVL${level}`;
  }
}

// parse a cec-client-style byte string, "1F:82:10:00" or "1f 82 10 00", into an
// array of bytes.
function parseBytes(text) {
  return text
    .trim()
    .split(/[\s:]+/)
    .filter((s) => s.length)
    .map((s) => parseInt(s, 16) & 0xff);
}

// ---------------------------------------------------------------------------
// main
// ---------------------------------------------------------------------------

function printHelp() {
  process.stdout.write(
`Interactive commands:
[tx] {bytes}              transfer bytes over the CEC line (e.g. tx 1F:82:10:00).
[txn] {bytes}             transfer bytes but don't wait for an ACK.
[on] {address}            power on the device with the given logical address.
[standby] {address}       put the device with the given address in standby.
[as]                      make the CEC adapter the active source.
[is]                      mark the CEC adapter as inactive source.
[sp] {physical address}   make the given physical address (e.g. 1.0.0.0) active.
[spl] {address}           make the given logical address active.
[osd] {addr} {string}     set an OSD message on the specified device.
[ver] {addr}              get the CEC version of the specified device.
[ven] {addr}              get the vendor id of the specified device.
[pow] {addr}              get the power status of the specified device.
[name] {addr}             get the OSD name of the specified device.
[poll] {addr}             poll the specified device.
[lad]                     list the active devices on the bus.
[scan]                    scan the CEC bus and display device info.
[self]                    show the logical addresses controlled by libCEC.
[volup]                   send a volume up command to the amp, if present.
[voldown]                 send a volume down command to the amp, if present.
[mute]                    send a mute command to the amp, if present.
[mon] {1|0}               enable or disable CEC bus monitoring.
[ping]                    send a ping command to the CEC adapter.
[log] {1-31}              change the log-level bitmask. see cectypes.h.
[r]                       reconnect to the CEC adapter.
[h] or [help]             show this help.
[q] or [quit]             quit.
`);
}

function main() {
  const opts = parseArgs(process.argv.slice(2));

  if (opts.help) {
    process.stdout.write(`${require('../package.json').description}\n\n`);
    printHelp();
    return;
  }

  // A bare adapter, used for the connection-less commands (list/info) and as the
  // handle the REPL drives. Passing the event callbacks up front demonstrates the
  // EventEmitter surface; every one maps to an ICECCallbacks slot in the addon.
  let logMask = opts.logMask;
  const adapter = new cec.CecAdapter({
    deviceName: 'CECNode',
    deviceType: opts.deviceType,
    monitorOnly: opts.monitorOnly,
    ...(opts.hdmiPort != null ? { hdmiPort: opts.hdmiPort } : {}),
    ...(opts.baseDevice != null ? { baseDevice: opts.baseDevice } : {}),
  });

  adapter.on('log', (m) => {
    if (m.level & logMask)
      console.log(`${fmtLogLevel(m.level)}[${m.time}]\t${m.message}`);
  });
  adapter.on('keyPress', (k) => {
    // duration 0 is the initial press; a follow-up with a duration is the release.
    if (k.duration === 0)
      console.log(`key pressed: ${cec.userControlKeyToString(k.keycode)} (${k.keycode})`);
  });
  adapter.on('command', (c) => {
    console.log(`>> ${fmtAddr(c.initiator)} -> ${fmtAddr(c.destination)}: ` +
                `${cec.opcodeToString(c.opcode)}` +
                (c.parameters.length ? ` [${c.parameters.map((b) => b.toString(16)).join(' ')}]` : ''));
  });
  adapter.on('sourceActivated', (address, activated) => {
    console.log(`source ${activated ? 'activated' : 'deactivated'}: ${fmtAddr(address)}`);
  });
  adapter.on('alert', (a) => console.log(`ALERT: ${a}`));
  adapter.on('configurationChanged', (config) => {
    if (config.logicalAddresses.length)
      console.log(`configuration changed; controlling ${config.logicalAddresses.map(fmtAddr).join(', ')}`);
  });

  // ---- connection-less commands -------------------------------------------

  const adapters = adapter.detectAdapters();

  if (opts.info) {
    console.log(adapter.getLibInfo());
    adapter.close();
    return;
  }

  if (opts.listDevices) {
    console.log(`found ${adapters.length} adapter(s):`);
    adapters.forEach((d, i) => {
      console.log(`device:              ${i + 1}`);
      console.log(`com port:            ${d.comName}`);
      console.log(`path:                ${d.path}`);
      console.log(`vendor id:           ${d.vendorId.toString(16)}`);
      console.log(`product id:          ${d.productId.toString(16)}`);
      console.log(`firmware version:    ${d.firmwareVersion}`);
      console.log(`type:                ${adapterTypeName(d.type)}\n`);
    });
    adapter.close();
    return;
  }

  // ---- open the adapter ----------------------------------------------------

  // open by the /dev node (strComName), as cec-client does — not the sysfs path.
  const port = opts.port || (adapters[0] && adapters[0].comName);
  if (!port) {
    console.error('no CEC adapter found; specify a port or connect one');
    adapter.close();
    process.exit(1);
  }

  console.log(`opening ${port} ...`);
  if (!adapter.open(port)) {
    console.error(`could not open a connection to ${port}`);
    adapter.close();
    process.exit(1);
  }
  console.log(adapter.getLibInfo());
  console.log("connection opened. type 'h' for help, 'q' to quit.");

  // ---- the REPL ------------------------------------------------------------

  const rl = readline.createInterface({ input: process.stdin, output: process.stdout, prompt: '' });
  rl.prompt();

  // Each command below is one focused example of the API. `addr()` reads a
  // logical address argument, defaulting to the TV.
  function addr(args, fallback = cec.CecLogicalAddress.TV) {
    return args.length ? parseInt(args[0], 10) : fallback;
  }

  function handle(line) {
    const parts = line.trim().split(/\s+/);
    const cmd = parts.shift();
    if (!cmd) return;

    switch (cmd) {
      case 'tx':
      case 'txn': {
        // Build a cec_command from raw bytes: the high/low nibble of the first
        // byte are the initiator/destination, the second byte is the opcode, the
        // rest are parameters. 'txn' asks not to wait for the ACK.
        const bytes = parseBytes(parts.join(' '));
        if (!bytes.length) { console.log('usage: tx <bytes>, e.g. tx 1F:82:10:00'); break; }
        const command = {
          initiator: (bytes[0] >> 4) & 0xf,
          destination: bytes[0] & 0xf,
          opcode: bytes.length > 1 ? bytes[1] : cec.CecOpcode.None,
          parameters: bytes.slice(2),
          transmitTimeout: cmd === 'txn' ? 0 : 1000,
        };
        console.log(adapter.transmit(command) ? 'transmit succeeded' : 'transmit failed');
        break;
      }

      case 'on':
        console.log(adapter.powerOnDevices(addr(parts, cec.CecLogicalAddress.Broadcast)) ? 'ok' : 'failed');
        break;

      case 'standby':
        console.log(adapter.standbyDevices(addr(parts, cec.CecLogicalAddress.Broadcast)) ? 'ok' : 'failed');
        break;

      case 'as':
        console.log(adapter.setActiveSource() ? 'now the active source' : 'failed');
        break;

      case 'is':
        console.log(adapter.setInactiveView() ? 'marked inactive' : 'failed');
        break;

      case 'sp': {
        // accept either "1.0.0.0" or a raw hex like "1000"
        const t = parts[0] || '';
        const pa = t.includes('.')
          ? t.split('.').reduce((acc, n) => (acc << 4) | (parseInt(n, 10) & 0xf), 0)
          : parseInt(t, 16);
        console.log(adapter.setStreamPathPhysical(pa) ? `stream path -> ${fmtPhysical(pa)}` : 'failed');
        break;
      }

      case 'spl':
        console.log(adapter.setStreamPathLogical(addr(parts)) ? 'ok' : 'failed');
        break;

      case 'osd': {
        const a = parseInt(parts.shift(), 10);
        const message = parts.join(' ');
        console.log(adapter.setOSDString(a, cec.CecDisplayControl.DisplayForDefaultTime, message) ? 'ok' : 'failed');
        break;
      }

      case 'ver':
        console.log(`CEC version ${fmtAddr(addr(parts))}: ${cec.cecVersionToString(adapter.getDeviceCecVersion(addr(parts)))}`);
        break;

      case 'ven':
        console.log(`vendor ${fmtAddr(addr(parts))}: ${cec.vendorIdToString(adapter.getDeviceVendorId(addr(parts)))}`);
        break;

      case 'pow':
        console.log(`power status ${fmtAddr(addr(parts))}: ${cec.powerStatusToString(adapter.getDevicePowerStatus(addr(parts)))}`);
        break;

      case 'name':
        console.log(`OSD name ${fmtAddr(addr(parts))}: ${adapter.getDeviceOSDName(addr(parts))}`);
        break;

      case 'poll':
        console.log(adapter.pollDevice(addr(parts)) ? 'device responded' : 'no response');
        break;

      case 'lad': {
        const active = adapter.getActiveDevices();
        console.log(`active devices: ${active.map(fmtAddr).join(', ') || '(none)'}`);
        break;
      }

      case 'scan':
        scanBus(adapter);
        break;

      case 'self': {
        const la = adapter.getLogicalAddresses();
        console.log(`controlled by libCEC: ${la.addresses.map(fmtAddr).join(', ') || '(none)'} ` +
                    `(primary ${fmtAddr(la.primary)})`);
        break;
      }

      case 'volup': console.log(`volume: ${adapter.volumeUp()}`); break;
      case 'voldown': console.log(`volume: ${adapter.volumeDown()}`); break;
      case 'mute': console.log(`audio status: ${adapter.muteAudio()}`); break;

      case 'mon': {
        const enable = parts[0] !== '0';
        console.log(adapter.switchMonitoring(enable) ? `monitoring ${enable ? 'on' : 'off'}` : 'failed');
        break;
      }

      case 'ping':
        console.log(adapter.pingAdapters() ? 'ping ok' : 'ping failed');
        break;

      case 'log':
        logMask = parseInt(parts[0], 10) || logMask;
        console.log(`log-level mask is now ${logMask}`);
        break;

      case 'r':
        console.log('reconnecting ...');
        adapter.close();
        // a fresh connection re-runs address allocation; re-open the same port.
        if (adapter.open(port)) console.log('reconnected'); else console.log('reconnect failed');
        break;

      case 'h': case 'help':
        printHelp();
        break;

      case 'q': case 'quit':
        rl.close();
        return;

      default:
        console.log(`unknown command '${cmd}'. type 'h' for help.`);
    }
  }

  rl.on('line', (line) => { try { handle(line); } catch (e) { console.error(`error: ${e.message}`); } rl.prompt(); });
  rl.on('close', () => { console.log('closing ...'); adapter.close(); process.exit(0); });
}

// A full bus scan, mirroring cec-client's `scan`: walk the active devices and
// query each attribute individually — a good end-to-end exercise of the query API.
function scanBus(adapter) {
  console.log('requesting CEC bus information ...');
  const devices = adapter.getActiveDevices();
  for (const address of devices) {
    const vendor = adapter.getDeviceVendorId(address);
    const physical = adapter.getDevicePhysicalAddress(address);
    const cecVersion = adapter.getDeviceCecVersion(address);
    const power = adapter.getDevicePowerStatus(address);
    const osdName = adapter.getDeviceOSDName(address);
    const active = adapter.isActiveSource(address);

    console.log(`device #${address}: ${cec.logicalAddressToString(address)}`);
    console.log(`address:       ${fmtPhysical(physical)}`);
    console.log(`active source: ${active ? 'yes' : 'no'}`);
    console.log(`vendor:        ${cec.vendorIdToString(vendor)}`);
    console.log(`osd string:    ${osdName}`);
    console.log(`CEC version:   ${cec.cecVersionToString(cecVersion)}`);
    console.log(`power status:  ${cec.powerStatusToString(power)}`);
    console.log('');
  }
  console.log(`scan complete: ${devices.length} device(s).`);
}

main();

File: src/rust/examples/cec_client.rs
// This file is part of the libCEC(R) library.
//
// libCEC(R) is Copyright (C) 2011-2026 Pulse-Eight Limited.  All rights reserved.
// libCEC(R) is an original work, containing original code.
//
// libCEC(R) is a trademark of Pulse-Eight Limited.
//
// This program is dual-licensed; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//
//
// Alternatively, you can license this library under a commercial license,
// please contact Pulse-Eight Licensing for more information.
//
// For more information contact:
// Pulse-Eight Licensing       <license@pulse-eight.com>
//     http://www.pulse-eight.com/
//     http://www.pulse-eight.net/

//! An interactive CEC console, in the shape of `cec-client` and the Node
//! binding's `cec-client.js`. This is the manual smoke test: the repo has no
//! automated suite for CEC behaviour, because that needs real hardware on a
//! real HDMI bus.
//!
//!     cargo run --example cec_client -- --help
//!     cargo run --example cec_client -- -p 1
//!
//! Type `h` at the prompt for the command list.

use std::io::{self, BufRead, Write};
use std::sync::mpsc::Receiver;
use std::thread;
use std::time::Duration;

use libcec::callbacks::channel;
use libcec::enums::{DeviceType, DisplayControl, LogicalAddress, Opcode, UserControlCode};
use libcec::{format_physical_address, CecEvent, Command, Connection, ConnectionBuilder, Error};

struct Options {
    port: Option<String>,
    hdmi_port: Option<u8>,
    base_device: Option<LogicalAddress>,
    device_type: DeviceType,
    monitor_only: bool,
    list_devices: bool,
    info: bool,
    log_level: i32,
}

impl Default for Options {
    fn default() -> Self {
        Options {
            port: None,
            hdmi_port: None,
            base_device: None,
            device_type: DeviceType::RecordingDevice,
            monitor_only: false,
            list_devices: false,
            info: false,
            // Error | Warning | Notice - the same default cec-client uses.
            log_level: 7,
        }
    }
}

fn print_help() {
    println!(
        "\
usage: cec_client [options]

options:
  -p, --port <n>       HDMI port the adapter is plugged into (1-based)
  -b, --base <addr>    logical address of the device it is plugged into
  -t, --type <type>    tv | recording | tuner | playback | audio
  -s, --serial <port>  open this port instead of the first adapter found
  -m, --monitor        watch the bus without claiming an address
  -l, --list-devices   list adapters and exit
  -i, --info           print adapter info and exit
  -d, --log-level <n>  log bitmask: 1 error, 2 warning, 4 notice, 8 traffic, 16 debug
  -h, --help           this text

commands (type at the prompt):
  tx <bytes>       transmit raw bytes, e.g. tx 10:04
  on [addr]        power on a device (default: TV)
  standby [addr]   put a device in standby (default: TV)
  as               make this client the active source
  is               make this client inactive
  sp <phys>        set the stream path, e.g. sp 1.0.0.0
  spl <addr>       set the stream path to a logical address
  osd <addr> <msg> put a message on a device's display
  key <addr> <n>   send user control code n, then release
  volup/voldown    volume up/down on the amplifier
  mute             toggle mute on the amplifier
  ver <addr>       the CEC version a device speaks
  ven <addr>       a device's vendor id
  pow <addr>       a device's power status
  name <addr>      a device's OSD name
  lang <addr>      a device's menu language
  poll <addr>      poll an address
  scan             scan the bus and describe every device
  self             the addresses this client holds
  stats            adapter frame counters
  mon <0|1>        turn monitoring mode on or off
  ping             check the adapter is answering
  r                rescan the bus
  h, help          this text
  q, quit          exit"
    );
}

fn parse_address(text: &str) -> Option<LogicalAddress> {
    text.parse::<i32>().ok().map(LogicalAddress::from_raw)
}

/// `1.0.0.0` or `1000` -> 0x1000.
fn parse_physical(text: &str) -> Option<u16> {
    if text.contains('.') {
        let parts: Vec<&str> = text.split('.').collect();
        if parts.len() != 4 {
            return None;
        }
        let mut address = 0u16;
        for part in parts {
            let nibble = u16::from_str_radix(part, 16).ok()?;
            if nibble > 0xF {
                return None;
            }
            address = (address << 4) | nibble;
        }
        Some(address)
    } else {
        u16::from_str_radix(text, 16).ok()
    }
}

/// `10:04` or `10 04` -> [0x10, 0x04].
fn parse_bytes(text: &str) -> Option<Vec<u8>> {
    text.split([':', ' ', ','])
        .filter(|part| !part.is_empty())
        .map(|part| u8::from_str_radix(part, 16).ok())
        .collect()
}

fn parse_args() -> Result<Options, String> {
    let mut options = Options::default();
    let mut args = std::env::args().skip(1);
    while let Some(arg) = args.next() {
        let mut value = || args.next().ok_or(format!("{arg} needs a value"));
        match arg.as_str() {
            "-h" | "--help" => {
                print_help();
                std::process::exit(0);
            }
            "-p" | "--port" => options.hdmi_port = Some(value()?.parse().map_err(|_| "bad port")?),
            "-b" | "--base" => {
                options.base_device = Some(parse_address(&value()?).ok_or("bad address")?)
            }
            "-s" | "--serial" => options.port = Some(value()?),
            "-d" | "--log-level" => {
                options.log_level = value()?.parse().map_err(|_| "bad log level")?
            }
            "-t" | "--type" => {
                options.device_type = match value()?.as_str() {
                    "tv" => DeviceType::Tv,
                    "recording" => DeviceType::RecordingDevice,
                    "tuner" => DeviceType::Tuner,
                    "playback" => DeviceType::PlaybackDevice,
                    "audio" => DeviceType::AudioSystem,
                    other => return Err(format!("unknown device type {other}")),
                }
            }
            "-m" | "--monitor" => options.monitor_only = true,
            "-l" | "--list-devices" => options.list_devices = true,
            "-i" | "--info" => options.info = true,
            other => return Err(format!("unknown option {other}")),
        }
    }
    Ok(options)
}

/// Print events as they arrive, on a thread of our own so the prompt stays
/// responsive. This is the reason to prefer the channel form of the callbacks:
/// printing from libCEC's worker thread would hold up the bus.
fn watch(events: Receiver<CecEvent>, log_level: i32) {
    thread::spawn(move || {
        for event in events {
            match event {
                CecEvent::LogMessage(message) => {
                    if message.level.raw() & log_level != 0 {
                        println!("[{:?}] {}", message.level, message.message);
                    }
                }
                CecEvent::KeyPress(key) => println!("key: {key}"),
                CecEvent::Command(command) => println!("cmd: {command}"),
                CecEvent::Alert(alert) => println!("alert: {alert:?}"),
                CecEvent::SourceActivated { address, activated } => println!(
                    "source {address} {}",
                    if activated {
                        "activated"
                    } else {
                        "deactivated"
                    }
                ),
                CecEvent::ConfigurationChanged(config) => println!(
                    "configuration: physical address {}",
                    format_physical_address(config.physical_address)
                ),
                CecEvent::MenuStateChanged(state) => println!("menu: {state}"),
            }
        }
    });
}

fn scan_bus(cec: &Connection) {
    println!("scanning the bus ...");
    for address in cec.active_devices().addresses {
        println!("device #{}: {address}", address.raw());
        println!(
            "  address:       {}",
            format_physical_address(cec.device_physical_address(address))
        );
        println!("  active source: {}", cec.is_active_source(address));
        println!("  vendor:        {:06x}", cec.device_vendor_id(address));
        println!("  osd string:    {}", cec.device_osd_name(address));
        println!("  CEC version:   {}", cec.device_cec_version(address));
        println!("  power status:  {}", cec.power_status(address));
        if let Some(language) = cec.device_menu_language(address) {
            println!("  language:      {language}");
        }
        println!();
    }
}

fn run_command(cec: &Connection, line: &str) -> Result<bool, Error> {
    let mut parts = line.split_whitespace();
    let Some(command) = parts.next() else {
        return Ok(true);
    };
    let rest: Vec<&str> = parts.collect();
    let address = |index: usize, fallback: LogicalAddress| {
        rest.get(index)
            .and_then(|text| parse_address(text))
            .unwrap_or(fallback)
    };

    match command {
        "q" | "quit" => return Ok(false),
        "h" | "help" => print_help(),

        "tx" => match parse_bytes(&rest.join(" ")) {
            // A raw frame is initiator/destination in the first byte, opcode in
            // the second, parameters after that - the same shape cec-client's
            // tx takes.
            Some(bytes) if bytes.len() >= 2 => {
                let command = Command {
                    initiator: LogicalAddress::from_raw((bytes[0] >> 4) as i32),
                    destination: LogicalAddress::from_raw((bytes[0] & 0x0F) as i32),
                    opcode: Opcode::from_raw(bytes[1] as i32),
                    parameters: bytes[2..].to_vec(),
                    ..Command::new(LogicalAddress::Unknown, Opcode::None)
                };
                cec.transmit(&command)?;
                println!("sent {command}");
            }
            _ => println!("usage: tx <bytes>, e.g. tx 10:04"),
        },

        "on" => cec.power_on(address(0, LogicalAddress::Tv))?,
        "standby" => cec.standby(address(0, LogicalAddress::Tv))?,
        "as" => cec.set_active_source(DeviceType::RecordingDevice)?,
        "is" => cec.set_inactive_view()?,

        "sp" => match rest.first().and_then(|text| parse_physical(text)) {
            Some(physical) => cec.set_stream_path(physical)?,
            None => println!("usage: sp <physical address>, e.g. sp 1.0.0.0"),
        },
        "spl" => cec.set_stream_path_to(address(0, LogicalAddress::Tv))?,

        "osd" => {
            if rest.len() < 2 {
                println!("usage: osd <address> <message>");
            } else {
                cec.set_osd_string(
                    address(0, LogicalAddress::Tv),
                    DisplayControl::DisplayForDefaultTime,
                    &rest[1..].join(" "),
                )?;
            }
        }

        "key" => match rest.get(1).and_then(|text| text.parse::<i32>().ok()) {
            Some(code) => {
                let destination = address(0, LogicalAddress::Tv);
                let key = UserControlCode::from_raw(code);
                cec.send_keypress(destination, key, true)?;
                cec.send_key_release(destination, true)?;
                println!("sent {key}");
            }
            None => println!("usage: key <address> <user control code>"),
        },

        "volup" => println!("audio: {}", cec.volume_up(true)),
        "voldown" => println!("audio: {}", cec.volume_down(true)),
        "mute" => println!("audio: {}", cec.mute_audio(true)),

        "ver" => println!("{}", cec.device_cec_version(address(0, LogicalAddress::Tv))),
        "ven" => println!(
            "{:06x}",
            cec.device_vendor_id(address(0, LogicalAddress::Tv))
        ),
        "pow" => println!("{}", cec.power_status(address(0, LogicalAddress::Tv))),
        "name" => println!("{}", cec.device_osd_name(address(0, LogicalAddress::Tv))),
        "lang" => match cec.device_menu_language(address(0, LogicalAddress::Tv)) {
            Some(language) => println!("{language}"),
            None => println!("unknown"),
        },
        "poll" => println!(
            "{}",
            if cec.poll_device(address(0, LogicalAddress::Tv)) {
                "ok"
            } else {
                "no answer"
            }
        ),

        "scan" => scan_bus(cec),
        "self" => {
            let addresses = cec.logical_addresses();
            println!("primary: {}", addresses.primary);
            for address in addresses.addresses {
                println!("  {address}");
            }
        }
        "stats" => println!("{:?}", cec.stats()?),
        "mon" => cec.switch_monitoring(rest.first() != Some(&"0"))?,
        "ping" => {
            cec.ping_adapter()?;
            println!("ok");
        }
        "r" => {
            cec.rescan_devices();
            println!("rescanned");
        }

        other => println!("unknown command '{other}' - type h for help"),
    }
    Ok(true)
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let options = match parse_args() {
        Ok(options) => options,
        Err(message) => {
            eprintln!("error: {message}");
            eprintln!("try --help");
            std::process::exit(2);
        }
    };

    if options.list_devices {
        for adapter in Connection::detect_adapters(false)? {
            println!("{adapter}");
            println!("  path:     {}", adapter.path);
            println!("  vendor:   {:04x}", adapter.vendor_id);
            println!("  product:  {:04x}", adapter.product_id);
            println!("  firmware: {}", adapter.firmware_version);
        }
        return Ok(());
    }

    let (handler, events) = channel();
    let mut builder = ConnectionBuilder::new("RustCEC")
        .device_type(options.device_type)
        .monitor_only(options.monitor_only)
        // Opening a console should not change what the television is showing.
        .activate_source(false)
        .callbacks(handler);
    if let Some(port) = options.hdmi_port {
        builder = builder.hdmi_port(port);
    }
    if let Some(base) = options.base_device {
        builder = builder.base_device(base);
    }

    watch(events, options.log_level);

    let cec = builder.open(options.port.as_deref(), Duration::from_secs(10))?;
    println!("{}", cec.lib_info());

    if options.info {
        println!("vendor:  {:04x}", cec.adapter_vendor_id());
        println!("product: {:04x}", cec.adapter_product_id());
        println!("{:?}", cec.configuration()?);
        return Ok(());
    }

    println!("type h for help, q to quit");
    let stdin = io::stdin();
    loop {
        print!("cec> ");
        io::stdout().flush()?;
        let mut line = String::new();
        if stdin.lock().read_line(&mut line)? == 0 {
            break; // EOF
        }
        // Strip a UTF-8 BOM: piping a script into this on Windows puts one on
        // the first line, and "﻿self" is a baffling way to be told a command
        // does not exist.
        let line = line.trim_start_matches('\u{feff}').trim();
        match run_command(&cec, line) {
            Ok(true) => {}
            Ok(false) => break,
            // A command that libCEC refused is not a reason to quit the console.
            Err(error) => println!("error: {error}"),
        }
    }

    println!("closing ...");
    Ok(())
}
